Fix part of #13822: Added backend validation for sub-topic url fragment

Lawful2002 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #13822.
2. This PR does the following: Adds backend validation to verify whether all subtopic url fragments are at most 25 chars in length and match the RegEx "^[a-z]+(-[a-z]+)*$"

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
![image](https://user-images.githubusercontent.com/73544247/172680015-77cf0cdf-a036-43d6-8c13-1628d0747c0f.png)
![image](https://user-images.githubusercontent.com/73544247/172680052-9cd29944-275c-4a7f-bc4b-07873540da80.png)


<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

Lawful2002 committed:
added beam job

Lawful2002 committed:
fixed typo

Lawful2002 committed:
fixed typo

Lawful2002 committed:
fixed tests

lkbhitesh07 commented:
Just a note @Lawful2002, Currently the frontend validation for the topic URL fragment is atmost 20 characters and here you are checking for 25 characters.
Thanks

Lawful2002 commented:
@lkbhitesh07 This [doc](https://docs.google.com/document/d/15M4F7-HIJltXUpE5XT6yimiR1pgr9bNdJRcqcEzGTso/edit) might need to be edited in that case. Thanks for noticing this.

vojtechjelinek commented on core/jobs/batch_jobs/topic_validation_jobs.py:
```diff
@@ -0,0 +1,104 @@
+        if url_frag is None:
+            return False
+
+        if len(url_frag) > 25:
```
Should this be 20?

Lawful2002 replied:
It should be 25. Eric has confirmed it

vojtechjelinek commented on core/jobs/batch_jobs/topic_validation_jobs.py:
```diff
@@ -0,0 +1,104 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Validation jobs for topic models"""
```
```suggestion
"""Validation jobs for topic models."""
```

Lawful2002 replied:
Fixed.

vojtechjelinek commented on core/jobs/batch_jobs/topic_validation_jobs.py:
```diff
@@ -0,0 +1,104 @@
+        if len(url_frag) > 25:
+            return False
+
+        regex = '^[a-z]+(-[a-z]+)*$'
```
From where did you got this regex?

Lawful2002 replied:
The regex was mentioned in this [doc](https://docs.google.com/document/d/15M4F7-HIJltXUpE5XT6yimiR1pgr9bNdJRcqcEzGTso/edit?usp=sharing)

vojtechjelinek commented on core/jobs/batch_jobs/topic_validation_jobs.py:
```diff
@@ -0,0 +1,104 @@
+
+        result = re.match(regex, url_frag)
+
+        return bool(result)
```
This can be done in one line IMO

Lawful2002 replied:
Done!

vojtechjelinek commented on core/jobs/batch_jobs/topic_validation_jobs_test.py:
```diff
@@ -0,0 +1,126 @@
+    TOPIC_ID_2 = '2'
+    TOPIC_ID_3 = '3'
+
+    def setUp(self):
```
Should we test URL fragment being `None`?

Lawful2002 replied:
I've added the test for the 'None' case in topic 3

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

U8NWXD commented:
Letting @vojtechjelinek take the first pass

Lawful2002 committed:
fixed job

Lawful2002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into subtopic-beam-job

Lawful2002 committed:
fixed linting

Lawful2002 committed:
addressed comments

Lawful2002 commented:
@vojtechjelinek I have updated this PR as I initially misunderstood the job. The job needs to verify the subtopic URL frag. instead of the topic URL frag. PTAL. (I have updated the sreenshots)

Lawful2002 committed:
Merge branch 'oppia:develop' into subtopic-beam-job

Lawful2002 commented:
@vojtechjelinek PTAL.

Lawful2002 committed:
Merge branch 'oppia:develop' into subtopic-beam-job

vojtechjelinek commented on core/jobs/batch_jobs/topic_validation_jobs.py:
```diff
@@ -0,0 +1,105 @@
+                lambda topic: (topic.id, list(topic.subtopics))
+            )
+            | 'Group subtopic info with topic id' >> beam.FlatMap(
+                lambda topic: [
+                    (topic[0], x['id'], x['url_fragment']) for x in topic[1]
+                    ]
```
```suggestion
                lambda topic: [
                    (topic[0], x['id'], x['url_fragment']) for x in topic[1]
                ]
```

Lawful2002 replied:
Done!

vojtechjelinek reviewed (changes requested):
Thanks! Only one nit feel free to submit testing request.

Lawful2002 committed:
Update core/jobs/batch_jobs/topic_validation_jobs.py

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

Lawful2002 commented:
@vojtechjelinek I have sent a job request. PTAL.

vojtechjelinek commented:
@Lawful2002 Hyy this passes with `TOPICS SUCCESS: 47`. Feel free to recreate it into a validation.

Lawful2002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into subtopic-beam-job

Lawful2002 committed:
added backend validation

Lawful2002 commented:
@vojtechjelinek I have added the backend validation. PTAL.

Lawful2002 committed:
added default arg

Lawful2002 committed:
fixed linting

vojtechjelinek commented on core/domain/topic_domain.py:
```diff
@@ -423,18 +423,21 @@ def from_dict(cls, subtopic_dict: SubtopicDict) -> Subtopic:
 
     @classmethod
-    def create_default_subtopic(cls, subtopic_id: int, title: str) -> Subtopic:
+    def create_default_subtopic(
+        cls, subtopic_id: int, title: str,
+        url_frag: str = 'url-frag') -> Subtopic:
```
```suggestion
    def create_default_subtopic(
        cls, 
        subtopic_id: int, 
        title: str,
        url_frag: str = 'url-frag'
    ) -> Subtopic:
```

Not sure if this is the correct solution.

Lawful2002 replied:
Done. PTAL.

vojtechjelinek commented on core/domain/topic_domain.py:
```diff
@@ -492,6 +495,17 @@ def validate(self) -> None:
                 'Expected subtopic title to be less than %d characters, '
                 'received %s' % (title_limit, self.title))
 
+        url_fragment_limit = android_validation_constants.MAX_CHARS_IN_SUBTOPIC_URL_FRAGMENT  # pylint: disable=line-too-long
```
```suggestion
        url_fragment_limit = (
            android_validation_constants.MAX_CHARS_IN_SUBTOPIC_URL_FRAGMENT)
```

vojtechjelinek replied:
Not resolved.

vojtechjelinek commented on core/domain/topic_domain_test.py:
```diff
@@ -491,6 +492,17 @@ def test_subtopic_title_validation(self) -> None:
+        self._assert_validation_error(
+            'Expected subtopic url fragment to be less '
+            'than or equal to %d characters' %
+            (android_validation_constants.MAX_CHARS_IN_SUBTOPIC_URL_FRAGMENT))
```
```suggestion
            android_validation_constants.MAX_CHARS_IN_SUBTOPIC_URL_FRAGMENT)
```

vojtechjelinek commented on core/domain/topic_domain.py:
```diff
@@ -492,6 +495,17 @@ def validate(self) -> None:
+                'than or equal to %d characters, received %s'
+                % (url_fragment_limit, self.url_fragment))
+
+        if not bool(re.match('^[a-z]+(-[a-z]+)*$', self.url_fragment)):
```
Should the regex be added to some constants?

Lawful2002 replied:
Done. I have added it under android constants.

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

U8NWXD commented:
Review by automated QA no longer required

Lawful2002 committed:
Update core/domain/topic_domain.py

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

Lawful2002 committed:
Update core/domain/topic_domain_test.py

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

Lawful2002 committed:
addressed comments

Lawful2002 committed:
addressed comments

Lawful2002 committed:
fixes

vojtechjelinek commented on core/controllers/admin.py:
```diff
@@ -471,7 +471,7 @@ def _load_dummy_new_structures_data(self):
             topic_1.add_uncategorized_skill_id(skill_id_2)
             topic_1.add_uncategorized_skill_id(skill_id_3)
-            topic_1.add_subtopic(1, 'Dummy Subtopic Title')
+            topic_1.add_subtopic(1, 'Dummy Subtopic Title', 'url-frag')
```
```suggestion
            topic_1.add_subtopic(1, 'Dummy Subtopic Title', 'dummysubtopic')
```

vojtechjelinek replied:
Lets change it here as this is not a test files an there might be a conflict if we generate this on the local server.

vojtechjelinek commented on core/controllers/android_e2e_config.py:
```diff
@@ -130,7 +130,7 @@ def post(self):
         topic.add_canonical_story(story_id)
         topic.add_uncategorized_skill_id(skill_id)
-        topic.add_subtopic(1, 'Test Subtopic Title')
+        topic.add_subtopic(1, 'Test Subtopic Title', 'url-frag')
```
```suggestion
        topic.add_subtopic(1, 'Test Subtopic Title', 'testsubtop')
```

ditto as in admin.py

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

Lawful2002 committed:
Update core/domain/topic_domain.py

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

Lawful2002 committed:
Update core/controllers/admin.py

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

Lawful2002 committed:
Update core/controllers/android_e2e_config.py

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

Lawful2002 committed:
fixed backend tests

Lawful2002 committed:
Merge remote-tracking branch 'origin/subtopic-beam-job' into subtopic-beam-job

Lawful2002 committed:
Merge branch 'oppia:develop' into subtopic-beam-job

Lawful2002 committed:
fixed tests

Lawful2002 committed:
fixed tests

Lawful2002 committed:
fixed backend tests

Lawful2002 committed:
fixed imdex.yaml

Lawful2002 committed:
fixed backend tests

Lawful2002 committed:
fixed backend tests

Lawful2002 commented:
@vojtechjelinek PTAL.

vojtechjelinek commented on core/domain/topic_domain.py:
```diff
@@ -149,7 +149,7 @@ class TopicChange(change_domain.BaseChange):
     }, {
         'name': CMD_ADD_SUBTOPIC,
-        'required_attribute_names': ['title', 'subtopic_id'],
+        'required_attribute_names': ['title', 'subtopic_id', 'url_fragment'],
```
Why is this added?

Lawful2002 replied:
This is added because the with the new validation, an empty `url_fragment` is invalid for the subtopic. The add_subtopic, create_default_subtopic functions both need to have a `url_fragment` arg, therefore the `ADD_SUBTOPIC` cmd also needs to have that attribute.

![image](https://user-images.githubusercontent.com/73544247/175980473-af1d7fb0-6503-42de-b737-059238a14b0f.png)

The frontend changes are required because we have updated the `ADD_SUBTOPIC` cmd, which is used in the frontend services.

vojtechjelinek reviewed (changes requested):
Thanks! One question? Also, why are the changes to the frontend needed?

Lawful2002 commented:
@vojtechjelinek PTAL.

seanlip reviewed (approved):
LGTM for codeowners :)

vojtechjelinek reviewed (approved):
LGTM!

Lawful2002 commented:
@gp201 @aks681 @kevintab95 PTAL.

Lawful2002 committed:
Merge branch 'oppia:develop' into subtopic-beam-job

Lawful2002 committed:
Merge branch 'oppia:develop' into subtopic-beam-job

gp201 reviewed (approved):
LGTM

Lawful2002 committed:
Merge branch 'oppia:develop' into subtopic-beam-job

Lawful2002 commented:
@kevintab95 PTAL.

kevintab95 reviewed (approved):
LGTM for codeowner files. Thanks @Lawful2002!
